test/params_api_test.c

```diff
@@ -0,0 +1,362 @@
+}
+
+static int test_param_construct(void)
+{
+    static const char *number_names[] = {
+        "int", "uint", "long", "ulong", "int32", "uint32",
+        "int64", "uint64", "size_t"
+    };
+    static const unsigned char bn_val[16] = {
+        0xac, 0x75, 0x22, 0x7d, 0x81, 0x06, 0x7a, 0x23,
+        0xa6, 0xed, 0x87, 0xc7, 0xab, 0xf4, 0x73, 0x22
+    };
+    OSSL_PARAM params[20];
+    char buf[100], buf2[100], *bufp, *bufp2;
+    unsigned char ubuf[100];
+    void *vp, *vp2;
+    const OSSL_PARAM *p;
+    static const OSSL_PARAM pend = OSSL_PARAM_END;
+    int i, n = 0, ret = 0;
+    unsigned int u;
+    long int l;
+    unsigned long int ul;
+    int32_t i32;
+    uint32_t u32;
+    int64_t i64;
+    uint64_t u64;
+    size_t j, k, s, sz;
+    double d, d2;
+    BIGNUM *bn = NULL, *bn2 = NULL;
+
+    params[n++] = OSSL_PARAM_construct_int("int", &i, &sz);
+    params[n++] = OSSL_PARAM_construct_uint("uint", &u, &sz);
+    params[n++] = OSSL_PARAM_construct_long("long", &l, &sz);
+    params[n++] = OSSL_PARAM_construct_ulong("ulong", &ul, &sz);
+    params[n++] = OSSL_PARAM_construct_int32("int32", &i32, &sz);
+    params[n++] = OSSL_PARAM_construct_int64("int64", &i64, &sz);
+    params[n++] = OSSL_PARAM_construct_uint32("uint32", &u32, &sz);
+    params[n++] = OSSL_PARAM_construct_uint64("uint64", &u64, &sz);
+    params[n++] = OSSL_PARAM_construct_size_t("size_t", &s, &sz);
+    params[n++] = OSSL_PARAM_construct_double("double", &d, &sz);
+    params[n++] = OSSL_PARAM_construct_BN("bignum", ubuf, sizeof(ubuf), &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_string("utf8str", buf, sizeof(buf),
+                                                   &sz);
+    params[n++] = OSSL_PARAM_construct_octet_string("octstr", buf, sizeof(buf),
+                                                    &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_ptr("utf8ptr", &bufp, &sz);
+    params[n++] = OSSL_PARAM_construct_octet_ptr("octptr", &vp, &sz);
+    memcpy(params + n, &pend, sizeof(params[n]));
+
+    /* Search failure */
+    if (!TEST_ptr_null(OSSL_PARAM_locate(params, "fnord")))
+        goto err;
+
+    /* All integral types */
+    for (j = 0; j < OSSL_NELEM(number_names); j++) {
+        if (!TEST_ptr(p = OSSL_PARAM_locate(params, number_names[j]))
+            || !TEST_true(OSSL_PARAM_set_uint32(p, (uint32_t)(3 + j)))
+            || !TEST_true(OSSL_PARAM_get_uint64(p, &u64))
+            || !TEST_size_t_eq(p->buffer_size, sz)
+            || !TEST_size_t_eq((size_t)u64, 3 + j)
+            || !TEST_true(OSSL_PARAM_get_double(p, &d2))
+            || !TEST_double_eq(d2, (double)u64)) {
+            TEST_note("iteration %zu var %s", j + 1, number_names[j]);
+            goto err;
+        }
+    }
+    /* Real */
+    if (!TEST_ptr(p = OSSL_PARAM_locate(params, "double"))
+        || !TEST_true(OSSL_PARAM_set_double(p, 3.14))
+        || !TEST_true(OSSL_PARAM_get_double(p, &d2))
+        || !TEST_size_t_eq(p->buffer_size, sz)
```

Comment:
Something more real world would be to `TEST_size_t_eq(sz, sizeof(double))`

Reply:
Changed.